### PR TITLE
Implement java.util.Properties.{load,store}.

### PR DESCRIPTION
**Pull request by ekrich**
So far `load` methods work but `store` do not.

Reference: https://github.com/scala-native/scala-native/pull/1100

### CODE REVIEW

**Comment by ekrich** [comment]
Well, the algorithm for `load` is not good. It is very slow and must produce a bunch of garbage because `boehm` GC in CI died for every combination. Both `immix` and `commix` passed but that doesn't excuse the performance. Here is the error form CI:
```
[info]   [ok] load(InputStream) with null input
[error] GC Warning: Failed to expand heap by -2146553856 bytes
[error] GC Warning: Failed to expand heap by -2147479552 bytes
[error] GC Warning: Out of Memory! Heap size: 99 MiB. Returning NULL!
java.net.SocketException: Socket is closed
```

**Comment by coreyoconnor** [comment]
From chat: The issue is with the handling of `mark` and how the native allocators handle overcommit. The current `BufferedInputStream` is from scalajs which is fine with overcommit, thus allocating a `MAX_INT` sized array does not immediately fail. However, at least `boehm`, will actually try to allocate memory of size `MAX_INT + overhead` . Which fails.

The proposed solution is to implement `mark` in the same way as harmony does: Allocates an initially small sized array and resizes this as necessary.

In short: have to fix `BufferedInputStream` first, then this will be fine :)

**Comment by ekrich** [comment]
Wow, thanks for looking into this.

**Comment by ekrich** [comment]
Since we are only marking the stream to look for an EBCDIC file byte, we don't need a high value for look ahead and reset. `mark` will allow `reset` up to the default buffer size of `8192` with any value less than or equal to `8192`. Any value higher than that will create a new buffer to the size specified. It was set to `Int.MaxValue` which is a big buffer for parsing a properties file.

**Comment by ekrich** [comment]
@sjrd Ready for review. This could also be ported to Scala.js.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 1]
```diff
@@ -67,19 +101,325 @@ class Properties(protected val defaults: Properties)
...
+    var unicode        = 0
+    var count          = 0
+    var nextChar: Char = 0
+    var buf            = new Array[Char](80)
+    var offset         = 0
+    var keyLength      = -1
+    val br             = new BufferedReader(reader)
+
+    @tailrec def processNext(isFirstChar: Boolean): Unit = {
+      val intVal = br.read()
```
Comment (sjrd) [inline javalib/src/main/scala/java/util/Properties.scala:186]: Consider using `br.readLine()`, then work on that string. `br.readLine()` already takes care of handling `\r`, `\n` and `\r\n`, so you don't have to mess with `isFirstChar` and `CONTINUE`. You would only have to parse line by line.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 2]
```diff
@@ -67,19 +101,325 @@ class Properties(protected val defaults: Properties)
...
+    var mode           = NONE
+    var unicode        = 0
+    var count          = 0
+    var nextChar: Char = 0
+    var buf            = new Array[Char](80)
+    var offset         = 0
+    var keyLength      = -1
+    val br             = new BufferedReader(reader)
+
+    @tailrec def processNext(isFirstChar: Boolean): Unit = {
```
Comment (sjrd) [inline javalib/src/main/scala/java/util/Properties.scala:185]: Please avoid pseudo-tailrec functions that actually modify surrounding state. That's twice as hard to understand, because one has to reason about the state *plus* the accumulators like `isFirstChar`. Prefer using a good old `while` loop instead.

Comment (sjrd) [reply javalib/src/main/scala/java/util/Properties.scala:185]: Overall, I find this algorithm that revolves around reading one character per loop iteration, and fiddle with a `mode` to save the state, *very* hard to reason about.

Have you considered using a much more straightforward approach that parses logical bits instead? Something with the following structure:
```scala
while (...) {
  val line = br.readLine()
  var i: Int = 0

  def parseUnicodeEscape(): Char = ...

  def parseKey(): String = {
    while (line.charAt(i) != '=') { ... }
  }

  def parseValue(): String = {
    while (i < line.length) { ... }
  }

  val key = parseKey()
  i += 1 // the '='
  val value = parseValue()

  putProperty(key, value)
}
```
Wouldn't that be so much easier to reason about?

File: javalib/src/main/scala/java/util/Properties.scala [hunk 3]
```diff
@@ -1,10 +1,26 @@
+/**
+ * Ported from Harmony
+ */
 package java.util
 
-import java.io.{PrintStream, PrintWriter}
+import java.io.{
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:6]: Thanks for the review, I know this code is a mess as it is pretty much a straight translation from Harmony and picked up from others. I took care of this one.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 4]
```diff
@@ -1,10 +1,26 @@
...
+  Reader,
+  Writer
+}
 import java.{util => ju}
 
+import scala.annotation.{switch, tailrec}
+import scala.collection.immutable.{Map => SMap}
 import scala.collection.JavaConverters._
-import scala.scalanative.annotation.stub
+import scala.util.control.Breaks._
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:23]: Yes, it was thanks.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 5]
```diff
@@ -14,11 +30,29 @@ class Properties(protected val defaults: Properties)
   def setProperty(key: String, value: String): AnyRef =
     put(key, value)
 
-  @stub
-  def load(inStream: java.io.InputStream): Unit = ???
+  def load(inStream: InputStream): Unit = {
+    if (inStream == null) {
+      throw new NullPointerException()
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:35]: Especially with no message exception. Done.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 6]
```diff
@@ -14,11 +30,29 @@ class Properties(protected val defaults: Properties)
...
+    }
+    val stream = new BufferedInputStream(inStream)
+    stream.mark(8192) // default buffer size
 
-  @stub
-  def load(reader: java.io.Reader): Unit = ???
+    val _isEbcdic = isEbcdic(stream)
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:40]: No this is fine. I think EBCDIC is here for Java on mainframe since this came from Harmony.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 7]
```diff
@@ -14,11 +30,29 @@ class Properties(protected val defaults: Properties)
...
+      loadImpl(new InputStreamReader(stream, "ISO8859-1"))
+    } else {
+      loadImpl(new InputStreamReader(stream))
+    }
+  }
+
+  def load(reader: Reader): Unit = {
+    if (reader == null) {
+      throw new NullPointerException()
+    }
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:53]: Yes, these are now removed.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 8]
```diff
@@ -14,11 +30,29 @@ class Properties(protected val defaults: Properties)
...
-  @stub
-  def load(reader: java.io.Reader): Unit = ???
+    val _isEbcdic = isEbcdic(stream)
+    stream.reset()
+
+    if (!_isEbcdic) {
+      loadImpl(new InputStreamReader(stream, "ISO8859-1"))
+    } else {
+      loadImpl(new InputStreamReader(stream))
+    }
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:47]: Yes, I think we should just remove it.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 9]
```diff
@@ -67,19 +101,325 @@ class Properties(protected val defaults: Properties)
   }
 
   def list(out: PrintStream): Unit = {
-    out.println("-- listing properties --")
-    entrySet().asScala.foreach { entry => out.println(format(entry)) }
+    list(new PrintWriter(new OutputStreamWriter(out), true))
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:104]: Yes, I reverted and just factored out the header string.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 10]
```diff
@@ -67,19 +101,325 @@ class Properties(protected val defaults: Properties)
...
   def list(out: PrintWriter): Unit = {
     out.println("-- listing properties --")
     entrySet().asScala.foreach { entry => out.println(format(entry)) }
   }
 
+  def store(out: OutputStream, comments: String): Unit = {
+    if (out == null) {
+      throw new NullPointerException()
+    }
+    // https://docs.oracle.com/javase/8/docs/api/java/util/Properties.html#load-java.io.InputStream-
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:116]: Removed, thanks.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 11]
```diff
@@ -67,19 +101,325 @@ class Properties(protected val defaults: Properties)
...
+
+    if (comments != null) {
+      writeComments(writer, comments)
+    }
+
+    writer.write('#')
+    writer.write(new Date().toString)
+    writer.write(System.lineSeparator)
+
+    val buffer = new StringBuilder(200)
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:134]: This was done to reuse the writer but that is ugly and a premature optimization. I moved the `StringBuilder` inside the `dumpString` which is also renamed to `encodeString`.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 12]
```diff
@@ -67,19 +101,325 @@ class Properties(protected val defaults: Properties)
...
+    var copyOfCh = ch
+    do {
+      hexChar = copyOfCh & 15
+      if (hexChar > 9) hexChar = hexChar - 10 + 'A'
+      else hexChar += '0'
+      index -= 1
+      hexChars(index) = hexChar.toChar
+    } while ({ copyOfCh >>>= 4; copyOfCh != 0 })
+    hexChars
+  }
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:420]: That is very pretty thank-you. I would have been hard pressed to come up with that but I can see the translation.

File: javalib/src/main/scala/java/util/Properties.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:40]: I just removed it.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 13]
```diff
@@ -67,19 +101,325 @@ class Properties(protected val defaults: Properties)
...
+        case '\b' =>
+          // On JVM \b get printed like \u0008 - See the following:
+          // https://docs.oracle.com/javase/8/docs/api/java/util/Properties.html#load-java.io.Reader-
+          buffer.appendAll(unicodeToHexaDecimal(ch))
+        case _ =>
+          if ("\\#!=:".indexOf(ch) >= 0 || (isKey && ch == ' '))
+            buffer.append('\\')
+          if (ch >= ' ' && ch <= '~') {
+            buffer.append(ch)
+          } else if (toHexaDecimal) {
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:397]: Yes, that is a good suggestion.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 14]
```diff
@@ -67,19 +101,325 @@ class Properties(protected val defaults: Properties)
...
     entrySet().asScala.foreach { entry => out.println(format(entry)) }
   }
 
+  def store(out: OutputStream, comments: String): Unit = {
+    if (out == null) {
+      throw new NullPointerException()
+    }
+    // https://docs.oracle.com/javase/8/docs/api/java/util/Properties.html#load-java.io.InputStream-
+    val writer = new OutputStreamWriter(out, "ISO8859_1")
+    store(writer, comments)
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:118]: I think the second commit addresses this so I will push it for review.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 15]
```diff
@@ -67,19 +101,325 @@ class Properties(protected val defaults: Properties)
...
+        case '\f' =>
+          buffer.append("\\f")
+        case '\r' =>
+          buffer.append("\\r")
+        case '\b' =>
+          // On JVM \b get printed like \u0008 - See the following:
+          // https://docs.oracle.com/javase/8/docs/api/java/util/Properties.html#load-java.io.Reader-
+          buffer.appendAll(unicodeToHexaDecimal(ch))
+        case _ =>
+          if ("\\#!=:".indexOf(ch) >= 0 || (isKey && ch == ' '))
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:354]: This seems good. I am not sure what we are matching with `\\`. Is that a backslash, is that the correct output?

[the remark this answers is not in the document]
Comment (sjrd) [reply javalib/src/main/scala/java/util/Properties.scala:354]: Yes, `'\\'` is 1 backslash, and it is output first as `'\\'` which is 1 backslash, followed by `ch` which is a second backslash.

[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:354]: The tests tell me you are correct - the code told me that the old way but I didn't trust it - to hard to read. Now it is much better.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 16]
```diff
@@ -67,19 +101,325 @@ class Properties(protected val defaults: Properties)
...
+        case '\n' =>
+          buffer.append("\\n")
+        case '\f' =>
+          buffer.append("\\f")
+        case '\r' =>
+          buffer.append("\\r")
+        case '\b' =>
+          // On JVM \b get printed like \u0008 - See the following:
+          // https://docs.oracle.com/javase/8/docs/api/java/util/Properties.html#load-java.io.Reader-
+          buffer.appendAll(unicodeToHexaDecimal(ch))
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala]: You are correct, thank-you.

File: javalib/src/main/scala/java/util/Properties.scala (hunk 2 above)
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:183]: I didn't write any of the current code but I knew it was super ugly. I couldn't reason about any of it honestly.

Thanks for the basic layout - I have made some progress. If I can get it to the point the tests pass then that would be great.

File: javalib/src/main/scala/java/util/Properties.scala (hunk 1 above)
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala]: I went with a line oriented approach - multi-line for continued values.

File: javalib/src/main/scala/java/util/Properties.scala (hunk 2 above)
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala]: I think the result is much more readable and maintainable. It has a very imperative style using `var`s and such.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 17]
```diff
@@ -67,19 +101,325 @@ class Properties(protected val defaults: Properties)
...
+      throw new NullPointerException()
+    }
+
+    if (comments != null) {
+      writeComments(writer, comments)
+    }
+
+    writer.write('#')
+    writer.write(new Date().toString)
+    writer.write(System.lineSeparator)
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:105]: I did the factoring.

**Comment by ekrich** [comment]
@sjrd When you get a chance this PR can be re-reviewed. Thanks in advance.

File: unit-tests/src/test/scala/java/util/PropertiesTest.scala [hunk 18]
```diff
@@ -0,0 +1,448 @@
...
+    def assertResult(result: String): Unit = {
+      val buffer4stream = new ByteArrayOutputStream
+      val stream        = new PrintStream(buffer4stream)
+      properties.list(stream)
+      assertEquals(buffer4stream.toString.trim, result.trim)
+      stream.flush()
+    }
+
+    val result0 = "-- listing properties --\n"
+    assertResult(result0)
```
Comment (sjrd) [inline unit-tests/src/test/scala/java/util/PropertiesTest.scala:201]: Any reason you're not directly passing the literal strings as arguments to `assertResult`?

File: javalib/src/main/scala/java/util/Properties.scala [hunk 19]
```diff
@@ -66,20 +74,304 @@ class Properties(protected val defaults: Properties)
...
+        char == ' ' || char == '\t' || char == '\f'
+
+      def isTokenKeySeparator(char: Char): Boolean =
+        char == '=' || char == ':'
+
+      def isKeySeparator(char: Char): Boolean =
+        isTokenKeySeparator(char) || isWhitespace(char)
+
+      def isEmpty(): Boolean =
+        line.isEmpty() // trim removes all whitespace
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:172]: This is an obsolete comment. Trim could not be used as trailing white space after line continuation is allowed.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 20]
```diff
@@ -66,20 +74,304 @@ class Properties(protected val defaults: Properties)
...
+      def valueContinues(): Boolean = oddBackslash()
+
+      def processChar(buf: jl.StringBuilder): Unit =
+        if (ch == '\\') {
+          ch = getNextChar
+          if (ch == 'u') {
+            getNextChar // advance
+            val uch = parseUnicodeEscape()
+            buf.append(uch)
+          } else if (ch == 't' || ch == 'f' || ch == 'r' || ch == 'n' || ch == 'b') {
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:197]: This looks much, much better and removes the map too. Thanks.

File: unit-tests/src/test/scala/java/util/PropertiesTest.scala (hunk 18 above)
Comment (ekrich) [reply unit-tests/src/test/scala/java/util/PropertiesTest.scala:201]: I don't see any reason so I changed it.

File: unit-tests/src/test/scala/java/util/PropertiesTest.scala [hunk 21]
```diff
@@ -0,0 +1,448 @@
...
+    assertEnumSameElementsAsSet[Any]("a", "b", "c")(prop.propertyNames())
+
+    val prop2 = new Properties(prop)
+    prop.setProperty("c", "CC")
+    prop.setProperty("d", "D")
+    assertEquals(4, enumerationSize(prop2.propertyNames()))
+    assertEnumSameElementsAsSet[Any]("a", "b", "c", "d")(prop2.propertyNames())
+  }
+
+  @Ignore @Test def propertyNamesIsNotAffectedByOverriddenPropertyNamesInDefaults()
```
[the remark this answers is not in the document]
Comment (ekrich) [reply unit-tests/src/test/scala/java/util/PropertiesTest.scala:74]: Something seems to be wrong with `default` handling - these tests are from Scala.js.

File: unit-tests/src/test/resources/properties-load-test.properties [hunk 22]
```diff
@@ -0,0 +1,34 @@
+# // Ported from Harmony
+
+
+
+
```
[the remark this answers is not in the document]
Comment (ekrich) [reply unit-tests/src/test/resources/properties-load-test.properties:5]: I looked and can't seem to find the original so I removed some to make it look more normal.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 23]
```diff
@@ -66,20 +74,304 @@ class Properties(protected val defaults: Properties)
...
+                           isKey: Boolean,
+                           toHex: Boolean): String = {
+    val buffer = new jl.StringBuilder(200)
+    var index  = 0
+    val length = string.length
+    // leading element (value) spaces are escaped
+    while (!isKey && index < length && string.charAt(index) == ' ') {
+      buffer.append("\\ ")
+      index += 1
+    }
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala:331]: Good idea. Thanks.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 24]
```diff
@@ -66,20 +74,304 @@ class Properties(protected val defaults: Properties)
...
+              && chars(indexPlusOne) == '\n') {
+            index += 1
+          }
+          writer.write(System.lineSeparator)
+          // return char with either '#' or '!' afterward
+          if (indexPlusOne < chars.length
+              && (chars(indexPlusOne) == '#'
+              || chars(indexPlusOne) == '!')) {
+            writer.write(chars(indexPlusOne))
+            index += 1
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala]: I reworked it to do this exactly.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 25]
```diff
@@ -66,20 +74,304 @@ class Properties(protected val defaults: Properties)
...
+          } else {
+            writer.write('#')
+          }
+
+        } else {
+          writer.write(chars(index))
+        }
+      } else {
+        if (toHex) {
+          writer.write(unicodeToHexaDecimal(chars(index)))
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala]: I think this is correct as this is just writing the comments. For the keys and values this is correct. I made a change to use the hex values so it matches the Javadoc. For comments it says "Latin 1" so I now use `ch >= 0xff`. I could reverse the logic to `> 0xff` for the encoding branch but it seems the normal case is without encoding but your call.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 26]
```diff
@@ -66,20 +74,304 @@ class Properties(protected val defaults: Properties)
       s"$key=$value"
   }
 
+  private val listStr = "-- listing properties --"
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala]: Done.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 27]
```diff
@@ -1,10 +1,16 @@
+/**
+ * Ported from Harmony
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala]: Not really so I removed it.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 28]
```diff
@@ -66,20 +74,304 @@ class Properties(protected val defaults: Properties)
...
+  }
+
+  private def writeComments(writer: Writer,
+                            comments: String,
+                            toHex: Boolean): Unit = {
+    writer.write('#')
+    val chars = comments.toCharArray
+    var index = 0
+    while (index < chars.length) {
+      if (chars(index) < 256) {
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala]: I couldn't get rid of all lookups but this is much better.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 29]
```diff
@@ -66,20 +74,304 @@ class Properties(protected val defaults: Properties)
...
 
   def list(out: PrintWriter): Unit = {
-    out.println("-- listing properties --")
+    out.println(listStr)
     entrySet().asScala.foreach { entry => out.println(format(entry)) }
   }
 
+  def store(out: OutputStream, comments: String): Unit = {
+    val writer = new OutputStreamWriter(out, StandardCharsets.ISO_8859_1)
+    storeImpl(writer, comments, true)
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala]: Specifying the parameter is much more readable. Ditto below.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 30]
```diff
@@ -66,20 +74,304 @@ class Properties(protected val defaults: Properties)
...
+  }
+
+  def store(writer: Writer, comments: String): Unit =
+    storeImpl(writer, comments, false)
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala]: Ditto.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 31]
```diff
@@ -66,20 +74,304 @@ class Properties(protected val defaults: Properties)
...
+
+    writer.write('#')
+    writer.write(new Date().toString)
+    writer.write(System.lineSeparator)
+
+    entrySet().asScala.foreach { entry =>
+      writer.write(encodeString(entry.getKey.asInstanceOf[String], true, toHex))
+      writer.write('=')
+      writer.write(
+        encodeString(entry.getValue.asInstanceOf[String], false, toHex))
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala]: Ditto.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 32]
```diff
@@ -66,20 +74,304 @@ class Properties(protected val defaults: Properties)
...
+    if (comments != null) {
+      writeComments(writer, comments, toHex)
+    }
+
+    writer.write('#')
+    writer.write(new Date().toString)
+    writer.write(System.lineSeparator)
+
+    entrySet().asScala.foreach { entry =>
+      writer.write(encodeString(entry.getKey.asInstanceOf[String], true, toHex))
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala]: Ditto.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 33]
```diff
@@ -66,20 +74,304 @@ class Properties(protected val defaults: Properties)
...
+          // return char with either '#' or '!' afterward
+          if (indexPlusOne < chars.length
+              && (chars(indexPlusOne) == '#'
+              || chars(indexPlusOne) == '!')) {
+            writer.write(chars(indexPlusOne))
+            index += 1
+          } else {
+            writer.write('#')
+          }
+
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala]: Removed thanks.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 34]
```diff
@@ -66,20 +74,304 @@ class Properties(protected val defaults: Properties)
...
+          }
+        }
+        valBuf.toString()
+      }
+
+      // run the parsing
+      if (!(isComment() || isEmpty())) {
+        ch = getNextChar
+        if (!isKeyParsed) {
+          valBuf = new jl.StringBuilder()
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala]: This is much better, thanks. I suppose the key buffer could be reused as well.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 35]
```diff
@@ -66,20 +74,304 @@ class Properties(protected val defaults: Properties)
...
+    var prevValueContinue = false
+    var isKeyParsed       = false
+    var key: String       = null
+    var line: String      = null
+
+    while ({ line = br.readLine(); line != null }) {
+      var i: Int   = -1
+      var ch: Char = Char.MinValue
+
+      def getNextChar: Char = {
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala]: Fixed.

File: javalib/src/main/scala/java/util/Properties.scala [hunk 36]
```diff
@@ -66,20 +74,304 @@ class Properties(protected val defaults: Properties)
...
+
+  @deprecated("", "")
+  def save(out: OutputStream, comments: String): Unit =
+    store(out, comments)
+
+  private def loadImpl(reader: Reader): Unit = {
+    import java.util.regex._
+    val trailingBackspace = Pattern.compile("""(\\)+$""")
+    lazy val chMap =
+      SMap('b' -> '\b', 'f' -> '\f', 'n' -> '\n', 'r' -> '\r', 't' -> '\t')
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/util/Properties.scala]: This is fixed and removed via the `match` suggestion you made.

File: unit-tests/src/test/scala/java/util/PropertiesTest.scala (hunk 21 above)
[the remark this answers is not in the document]
Comment (ekrich) [reply unit-tests/src/test/scala/java/util/PropertiesTest.scala:74]: These two tests do not pass, I overlooked these when porting from Scala.js. They have to do with the default properties. If you see something obvious let me know. I need to fix these two and remove `@ignore`.

[the remark this answers is not in the document]
Comment (ekrich) [reply unit-tests/src/test/scala/java/util/PropertiesTest.scala]: Resolved by using code from Scala.js and porting `ScalaOps` over.

File: unit-tests/src/test/scala/java/util/PropertiesTest.scala [hunk 37]
```diff
@@ -0,0 +1,448 @@
...
+    assertCollSameElementsAsSet("a", "b", "c")(prop.stringPropertyNames())
+
+    val prop2 = new Properties(prop)
+    prop.setProperty("c", "CC")
+    prop.setProperty("d", "D")
+    assertEquals(4, prop2.stringPropertyNames().size)
+    assertCollSameElementsAsSet("a", "b", "c", "d")(prop2.stringPropertyNames())
+  }
+
+  @Ignore @Test def stringPropertyNamesIsNotAffectedByOverriddenStringPropertyNamesInDefaults()
```
[the remark this answers is not in the document]
Comment (ekrich) [reply unit-tests/src/test/scala/java/util/PropertiesTest.scala]: Same as above, removed `@ignore`.

File: unit-tests/src/test/scala/java/util/PropertiesTest.scala (hunk 21 above)
[the remark this answers is not in the document]
Comment (sjrd) [reply unit-tests/src/test/scala/java/util/PropertiesTest.scala]: Ah, but all that new code has nothing to do with this PR :s

Could you set aside that last commit, remove it from this PR, and resubmit it as a separate PR once this one is merged please? And instead, *delete* the `@Ignored` tests in this PR, since they have nothing to do with `load` and `store` either. They should be part of the separate PR that focuses on fixing `propertyNames` and `stringPropertyNames`.

[the remark this answers is not in the document]
Comment (ekrich) [reply unit-tests/src/test/scala/java/util/PropertiesTest.scala]: Sure thing - Issue created: https://github.com/scala-native/scala-native/issues/1871
